### ISSUE TITLE
[mlir] "expected" diagnostics are checked even when `-verify-diagnostics` flag is not used

### ISSUE THREAD

**Issue by tim-hoffman** [llvm/llvm-project#222689]
Commit: 2e2c316fd5d7e8cff94703992928d5882d00478e

Input:
```mlir
// expected-error @unknown {{an error that is never produced}}
module {}
```

Command:
`build/bin/mlir-opt input.mlir`

Actual output:
```
module {
}

<unknown>:0: error: expected error "an error that is never produced" was not produced
```

Expected output:
```
module {
}
```

Relevant code:
https://github.com/llvm/llvm-project/blob/847fb5cfde68fdea490171bc24087ebb0cd90738/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp#L777-L796

The `SourceMgrDiagnosticVerifierHandler` is created unconditionally rather than guarded by `config.shouldVerifyDiagnostics()` and its destructor calls `verify()` which results in unsatisfied "expected" declarations being reported even when `config.shouldVerifyDiagnostics()` is false.
